packages/ast-spec/src/special/ExportSpecifier/spec.ts

```diff
@@ -2,9 +2,10 @@ import type { AST_NODE_TYPES } from '../../ast-node-types';
 import type { BaseNode } from '../../base/BaseNode';
 import type { ExportKind } from '../../declaration/ExportAndImportKind';
 import type { Identifier } from '../../expression/Identifier/spec';
+import type { StringLiteral } from '../../expression/literal/StringLiteral/spec';
 
 export interface ExportSpecifier extends BaseNode {
-  exported: Identifier;
+  exported: Identifier | StringLiteral;
   exportKind: ExportKind;
   local: Identifier;
```

Comment:
You have missed one case here

```suggestion
  local: Identifier | StringLiteral;
```

However note from [the estree spec](https://github.com/estree/estree/blob/3851d4a6eae5e5473371893959b88b62007469e8/es2022.md#exportspecifier)

> local can be Literal only if the source of the ExportNamedDeclaration of the parent of this node is not null. e.g. export { "foo" as "foo" } from "mod" is valid, export { "foo" as "foo" } is invalid.

So we should create two separate `ExportSpecifier`'s:
- One for the non-`source` export that only allows `local: Identifier`
- One for the `source` export that allows `local: Identifier | StringLiteral`

Reply:
I started on this in ced3289ca, but paused. These nodes don't have a `parent` in them. So is there really any difference on the `ast-spec` side?

Reply:
Yeah the difference is on the `ExportDeclaration` side of things where we can specify via the types and the conversion logic that an export without a source cannot have a specifier with a literal local.

Reply:
Hmm, I can't find any cases that TS doesn't directly error on. Am I missing something?

Reply:
You're likely getting **_SEMANTIC_** errors, not **_SYNTACTIC_** errors.

Syntactic errors are invariants and will crash out the TS parser -- i.e. we can just assume they'll crash and be happy.
Semantic errors will parse just fine and will not crash out the TS parser -- which means we have to throw our own errors.